### ISSUE TITLE
Epic: IP-whitelist for proxy

### ISSUE THREAD

**Issue by vadim2404** [neondatabase/neon#5628]
## Motivation
To forward traffic only from whitelisted hosts

## DoD
As a user, I'm able to configure IP whitelist
As a user, I'm unable to connect to my compute instance from non-whitelisted IP through pg protocol, WebSockets, and HTTP

## Implementation ideas
- Extend control plane API to expose the IP-ranges
- Forbid connections for endpoints from not listed IP address
- If the setting is not configured, it means, that all connections must be accepted

## Tasks
- [ ]

## Other related tasks and Epics
-

**Comment by ololobus** [comment]
From the cplane side list of allowed ips could be added into `ProxyWakeCompute` response
https://github.com/neondatabase/cloud/blob/535a96c67d63b1b108c76e353151af7cd336ab9b/openapispec/v2/management-v2.yaml#L904-L914

Like new `allowed_ips` arrays of strings

Go code of this handler is placed here
https://github.com/neondatabase/cloud/blob/535a96c67d63b1b108c76e353151af7cd336ab9b/goapp/internal/controllers/managementapiv2/authproxy.go#L105

Yet, we will also need to store the actual list somewhere in the control plane metadata. I tend to do not do this on the per branch / endpoint basis, but rather keep it project-scoped, then add endpoint-scope if users will ask about it. Then we can put `AllowedIPs` in to `DefaultEndpointSettings` in the project
https://github.com/neondatabase/cloud/blob/535a96c67d63b1b108c76e353151af7cd336ab9b/goapp/internal/dto/project.go#L252-L257

And also add it to public IP so that users was able to change it
https://github.com/neondatabase/cloud/blob/535a96c67d63b1b108c76e353151af7cd336ab9b/openapispec/v2/public-v2.yaml#L3880

And the same in the internal cplane API
https://github.com/neondatabase/cloud/blob/535a96c67d63b1b108c76e353151af7cd336ab9b/goapp/internal/controlplane/apispec/v1.yaml#L3875

**Comment by stepashka** [comment]
i'd like to flag addtional requirement, let's also add this to the issue body?
the user should be able to easily exclude the non-primary branches from the whitelisting
default to apply to all branches, but an option to only apply to the primary branch

hope this fits well into the scope of the same issue! @vadim2404 @khanova @ololobus

**Comment by vadim2404** [comment]
We discussed with @khanova where to put it and were thinking about `default_endpoint_settings`.

**Comment by stepashka** [comment]
@ololobus i suspect this will be hard to manage on per branch level for the user
i like your proposal for the bool flag for excluding non primary. 

@vadim2404 , not sure that `default_endpoint_settings` would do what's desired
let's catch up

**Comment by khanova** [comment]
@ololobus what do you think about adding it to `ProxyGetRoleSecret` instead? 
It, probably, makes more sense to be the part of authentication. Also if IP doesn't match, connection will be refused earlier.

**Comment by ololobus** [comment]
> what do you think about adding it to ProxyGetRoleSecret instead?
> It, probably, makes more sense to be the part of authentication.

Yes, it does, but iirc, proxy skips this call in several cases: when password hack is used https://neon.tech/docs/connect/connection-errors#d-specify-the-endpoint-id-in-the-password-field and in http and/or serverless mode. So we will whitelist some connection, but not all, which will be strange

@conradludgate can correct me, but I think that wake is the only common call for all connection options. Otherwise, yes it'd make more sense to check ip before waiking up

**Comment by conradludgate** [comment]
> proxy skips this call in several cases: when password hack is used

I'm addressing this at the moment, I'm going to make password hack not wake_compute without first authenticating

**Comment by vadim2404** [comment]
this week:

@khanova 
to merge control plane PR this week
to start on proxy part this week

**Comment by khanova** [comment]
- Project settings -> Default endpoint settings -> (allowlist + exclude non-primary branches)
- formats (v4-v6): single IP, mask
